src/webgl/3d_primitives.js

```diff
@@ -10,6 +10,179 @@ import p5 from '../core/main';
+  return this._renderer.beginGeometry();
+};
+
+/**
+ * Finishes creating a new <a href="#/p5.Geometry">p5.Geometry</a> that was
+ * started using <a href="#/p5/beginGeometry">beginGeometry()</a>. One can also
+ * use <a href="#/p5/buildGeometry">buildGeometry()</a> to pass a function that
+ * draws shapes.
+ *
+ * @method endGeometry
+ * @returns {p5.Geometry} The model that was built.
+ */
+p5.prototype.endGeometry = function() {
+  return this._renderer.endGeometry();
+};
+
+/**
+ * Creates a new <a href="#/p5.Geometry">p5.Geometry</a> that contains all
+ * the shapes drawn in a provided callback function. The returned combined shape
+ * can then be drawn all at once using <a href="#/p5/model">model()</a>.
+ *
+ * If you need to draw complex shapes every frame which don't change over time,
+ * combining them with `buildGeometry()` once and then drawing that will likely
+ * run faster than repeatedly drawing the individual pieces.
+ *
+ * One can also draw shapes directly between
+ * <a href="#/p5/beginGeometry">beginGeometry()</a> and
+ * <a href="#/p5/endGeometry">endGeometry()</a> instead of using a callback
+ * function.
+ *
+ * @method buildGeometry
+ * @param {Function} callback A function that draws shapes.
+ * @returns {p5.Geometry} The model that was built from the callback function.
+ *
+ * @example
+ * <div>
+ * <code>
+ * let tree;
+ * let button;
+ *
+ * function setup() {
+ *   createCanvas(100, 100, WEBGL);
+ *   button = createButton('New');
+ *   button.mousePressed(makeTree);
+ *   makeTree();
+ * }
+ *
+ * function makeTree() {
+ *   if (tree) freeGeometry(tree);
+ *   tree = buildGeometry(() => {
+ *     const addBranch = function(depth) {
+ *       push();
+ *       translate(0, -50);
+ *       cylinder(15, 100);
+ *       translate(0, -50);
+ *       if (depth >= 5) {
+ *         sphere(30);
+ *       } else {
+ *         const numChildren = round(random(1, 3));
+ *         for (let i = 0; i < numChildren; i++) {
+ *           push();
+ *           rotateZ(random(-0.3, 0.3) * PI);
+ *           rotateY(random(TWO_PI));
+ *           addBranch(depth + 1);
+ *           pop();
+ *         }
+ *       }
+ *       pop();
+ *     };
+ *     translate(0, 25);
+ *     scale(0.18);
+ *     addBranch(0);
+ *   });
+ * }
+ *
+ * function draw() {
+ *   background(255);
+ *   lights();
+ *   noStroke();
+ *   orbitControl();
+ *   model(tree);
+ * }
+ * </code>
+ * </div>
+ *
+ * @alt
+ * A 3D tree made of multiple cylinders and spheres.
+ */
+p5.prototype.buildGeometry = function(callback) {
+  return this._renderer.buildGeometry(callback);
+};
+
+/**
+ * Clears the resources of a model to free up browser memory. A model whose
+ * resources have been cleared can still be drawn, but the first time it is
+ * drawn again, it might take longer.
+ *
+ * @method freeGeometry
+ * @param {p5.Geometry} The geometry whose resources should be freed
+ */
+p5.prototype.freeGeometry = function(geometry) {
```

Comment:
if you call this on a model loaded with loadModel() what would happen?

Reply:
Same thing as with a combined model, so we clear the GPU resources associated with it. We still have a copy of all the data on the CPU so calling `model()` with it again will still work, it'll just be a bit slower the first time after we've cleared it

Reply:
Nice, that's what I would expect. Do you think it's worth noting this behavior?

Reply:
Yeah sounds good, it's easy to miss that `loadModel` also returns a `p5.Geometry` so it doesn't hurt to reiterate